### ISSUE TITLE
Gold Price Prediction & Risk Analysis

### ISSUE THREAD

**Issue by yoganithip24aid-lab** [yoganithip24aid-lab/my-project#1]
``"""
Enhanced Gold Price Analysis Mini-Project
With Machine Learning, Sentiment Analysis, and Advanced Features
"""

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import yfinance as yf
from datetime import datetime, timedelta
import warnings
warnings.filterwarnings('ignore')

from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

import plotly.graph_objects as go
from plotly.subplots import make_subplots
from scipy import stats

plt.style.use('seaborn-v0_8-darkgrid')
sns.set_palette("husl")

class EnhancedGoldAnalyzer:
    def __init__(self):
        self.gold_data = None
        self.macro_data = {}
        self.market_data = {}
        self.commodities_data = {}
        self.crypto_data = {}
        self.primary_gold = None
        self.model = None
        self.scaler = MinMaxScaler()

    def fetch_all_data(self, years=5):
        end_date = datetime.now()
        start_date = end_date - timedelta(days=365 * years)

        self.fetch_gold_data(start_date, end_date)
        self.fetch_macroeconomic_data(start_date, end_date)
        self.fetch_market_indices(start_date, end_date)
        self.fetch_commodities_data(start_date, end_date)
        self.fetch_crypto_data(start_date, end_date)

    def fetch_gold_data(self, start_date, end_date):
        gold_sources = {
            'Gold_Futures': 'GC=F',
            'GLD_ETF': 'GLD',
            'IAU_ETF': 'IAU'
        }

        self.gold_data = {}
        for name, ticker in gold_sources.items():
            data = yf.download(ticker, start=start_date, end=end_date, progress=False)
            if not data.empty:
                self.gold_data[name] = data[['Close']]

        if 'Gold_Futures' in self.gold_data:
            self.primary_gold = self.gold_data['Gold_Futures']
        elif 'GLD_ETF' in self.gold_data:
            self.primary_gold = self.gold_data['GLD_ETF']

    def fetch_macroeconomic_data(self, start_date, end_date):
        tickers = {
            'USD_Index': 'DX-Y.NYB',
            'VIX': '^VIX'
        }
        for name, ticker in tickers.items():
            data = yf.download(ticker, start=start_date, end=end_date, progress=False)
            if not data.empty:
                self.macro_data[name] = data[['Close']]

    def fetch_market_indices(self, start_date, end_date):
        indices = {
            'SP500': '^GSPC',
            'NASDAQ': '^IXIC'
        }
        for name, ticker in indices.items():
            data = yf.download(ticker, start=start_date, end=end_date, progress=False)
            if not data.empty:
                self.market_data[name] = data[['Close']]

    def fetch_commodities_data(self, start_date, end_date):
        commodities = {
            'Silver': 'SI=F',
            'Oil': 'CL=F'
        }
        for name, ticker in commodities.items():
            data = yf.download(ticker, start=start_date, end=end_date, progress=False)
            if not data.empty:
                self.commodities_data[name] = data[['Close']]

    def fetch_crypto_data(self, start_date, end_date):
        cryptos = {
            'Bitcoin': 'BTC-USD'
        }
        for name, ticker in cryptos.items():
            data = yf.download(ticker, start=start_date, end=end_date, progress=False)
            if not data.empty:
                self.crypto_data[name] = data[['Close']]

    def calculate_advanced_technical_indicators(self):
        if self.primary_gold is None:
            return None

        data = self.primary_gold.copy()
        data.columns = ['Close']

        data['Returns'] = data['Close'].pct_change()
        data['SMA_20'] = data['Close'].rolling(20).mean()
        data['SMA_50'] = data['Close'].rolling(50).mean()
        data['SMA_200'] = data['Close'].rolling(200).mean()

        ema12 = data['Close'].ewm(span=12).mean()
        ema26 = data['Close'].ewm(span=26).mean()
        data['MACD'] = ema12 - ema26
        data['MACD_Signal'] = data['MACD'].ewm(span=9).mean()

        delta = data['Close'].diff()
        gain = delta.clip(lower=0).rolling(14).mean()
        loss = -delta.clip(upper=0).rolling(14).mean()
        rs = gain / loss
        data['RSI'] = 100 - (100 / (1 + rs))

        data['BB_Middle'] = data['Close'].rolling(20).mean()
        std = data['Close'].rolling(20).std()
        data['BB_Upper'] = data['BB_Middle'] + 2 * std
        data['BB_Lower'] = data['BB_Middle'] - 2 * std
        data['BB_Position'] = (data['Close'] - data['BB_Lower']) / (data['BB_Upper'] - data['BB_Lower'])

        data['Volatility_20'] = data['Returns'].rolling(20).std() * np.sqrt(252)

        #  FIXED LINE (Pandas 2.x compatible)
        data = data.ffill().bfill()

        return data

    def create_interactive_dashboard(self):
        data = self.calculate_advanced_technical_indicators()
        if data is None:
            return

        fig = make_subplots(rows=2, cols=1, shared_xaxes=True)

        fig.add_trace(go.Scatter(x=data.index, y=data['Close'], name='Gold Price'), row=1, col=1)
        fig.add_trace(go.Scatter(x=data.index, y=data['SMA_50'], name='SMA 50'), row=1, col=1)
        fig.add_trace(go.Scatter(x=data.index, y=data['SMA_200'], name='SMA 200'), row=1, col=1)

        fig.add_trace(go.Scatter(x=data.index, y=data['RSI'], name='RSI'), row=2, col=1)

        fig.update_layout(height=900, title="Gold Price Interactive Dashboard")
        fig.write_html("gold_interactive_dashboard.html")

    def generate_comprehensive_report(self):
        data = self.calculate_advanced_technical_indicators()
        data.to_csv("gold_technical_data_enhanced.csv")

def main():
    print("ENHANCED GOLD PRICE ANALYSIS")

    analyzer = EnhancedGoldAnalyzer()
    analyzer.fetch_all_data(years=5)

    if analyzer.primary_gold is not None:
        analyzer.create_interactive_dashboard()
        analyzer.generate_comprehensive_report()
        print("ANALYSIS COMPLETE ")
    else:
        print("FAILED TO FETCH GOLD DATA ")

if __name__ == "__main__":
    main()
